internal/edittree/edittree.go

```diff
@@ -0,0 +1,878 @@
+		// original array. We build a new Array with modified/deleted keys.
+		out := make([]*ast.Term, 0, e.insertions.Length())
+		// Copy over all keys that weren't deleted/modified.
+		eIdx := 0
+		for i := 0; i < e.insertions.Length(); i++ {
+			if e.insertions.Element(i) == 0 {
+				// Scan through the `e.eliminated` bit-vec, and the first non-zero index, we pick.
+				foundIdx := false
+				for j := eIdx; j < e.eliminated.Length(); j++ {
+					if e.eliminated.Element(j) == 0 {
+						foundIdx = true
+						eIdx = j
+						break
+					}
+				}
+				if !foundIdx {
+					return nil, fmt.Errorf("too many eliminated indexes in array, expected to find uneliminated index %d", i)
+				}
+				// Append element from original term.
+				out = append(out, x.Elem(eIdx))
+				eIdx++ // Bump the counter, so that we monotonically advance through the array.
+			} else {
+				// Append value from rendered child index.
+				if t, ok := e.childScalarValues[i]; ok {
+					out = append(out, t)
+				} else if child, ok := e.childCompositeValues[i]; ok {
+					t, err := child.Render()
+					if err != nil {
+						return nil, err
+					}
+					out = append(out, t)
+				} else {
+					return nil, fmt.Errorf("invalid index %d does not exist in array", i)
+				}
+			}
+		}
+		return ast.ArrayTerm(out...), nil
+	default:
+		return e.value, nil
+	}
+}
+
+func (e *EditTree) InsertAtPath(path ast.Ref, value *ast.Term) (*EditTree, error) {
+	if value == nil {
+		return nil, fmt.Errorf("cannot insert nil value into EditTree")
+	}
+	if len(path) == 0 {
+		if value == nil {
+			return nil, fmt.Errorf("cannot insert nil value at root")
+		}
+		e.value = value
+		e.childKeys = map[int]*ast.Term{}
+		e.childScalarValues = map[int]*ast.Term{}
+		e.childCompositeValues = map[int]*EditTree{}
+		if v, ok := value.Value.(*ast.Array); ok {
+			length := ((v.Len() - 1) / 8) + 1
+			e.eliminated = bitvector.NewBitVector(make([]byte, length), v.Len())
+			e.insertions = bitvector.NewBitVector(make([]byte, length), v.Len())
+		}
+		return e, nil
+	}
+	dest, err := e.Unfold(path[:len(path)-1])
+	if err != nil {
+		return nil, err
+	}
+	return dest.Insert(path[len(path)-1], value)
+}
+
+func (e *EditTree) DeleteAtPath(path ast.Ref) (*EditTree, error) {
+	// Root document case:
+	if len(path) == 0 {
+		if e.value == nil {
+			return nil, fmt.Errorf("deleted node encountered during delete operation")
+		}
+		e.value = nil
+		e.childKeys = nil
+		e.childScalarValues = nil
+		e.childCompositeValues = nil
+		e.eliminated = nil
+		e.insertions = nil
+		return e, nil
+	}
+	dest, err := e.Unfold(path[:len(path)-1])
+	if err != nil {
+		return nil, err
+	}
+	return dest.Delete(path[len(path)-1])
+}
+
+func (e *EditTree) RenderAtPath(path ast.Ref) (*ast.Term, error) {
+	dest, err := e.Unfold(path)
+	if err != nil {
+		return nil, err
+	}
+	return dest.Render()
+}
+
+func (e *EditTree) String() string {
+	t, err := e.Render()
+	if err != nil {
+		return err.Error()
```

Comment:
[nit] I think it's OK to panic here. `fmt.Sprintf("%v", e)` will deal with that in a way.